src/main/java/com/onarandombox/MultiverseCore/listeners/MVPlayerListener.java

```diff
@@ -253,112 +243,140 @@ public void playerPortalCheck(PlayerPortalEvent event) {
+        if (MVPlayerListener.this.actionChecker.canKeepGameMode(player, world).isSuccessful()) {
+            Logging.fine("Player: " + player.getName() + " is IMMUNE to gamemode changes!");
+            return;
+        }
+
+        // Check that the player is in the new world and they haven't been teleported elsewhere or the event cancelled.
+        if (player.getWorld() != world.getCBWorld()) {
+            Logging.fine("The gamemode/allowfly was NOT changed for player '%s' because he is now in world '%s' instead of world '%s'",
+                    player.getName(), player.getWorld().getName(), world.getName());
+            return;
+        }
+        Logging.fine("Handling gamemode for player: %s, Changing to %s", player.getName(), world.getGameMode().toString());
+        Logging.finest("From World: %s", player.getWorld());
+        Logging.finest("To World: %s", world);
+        player.setGameMode(world.getGameMode());
+
+        // Check if their flight mode should change
+        // TODO need a override permission for this
+        if (player.getAllowFlight() && !world.getAllowFlight() && player.getGameMode() != GameMode.CREATIVE) {
+            player.setAllowFlight(false);
+            if (player.isFlying()) {
+                player.setFlying(false);
+            }
+        } else if (world.getAllowFlight()) {
+            if (player.getGameMode() == GameMode.CREATIVE) {
+                player.setAllowFlight(true);
+            }
+        }
+    }
+
+    /**
+     * Sends the reason for the teleportation failure to the sender.
+     *
+     * @param sender        The sender of the command.
+     * @param teleportee    The player being teleported.
+     * @param fromWorld     The world the player is teleporting from.
+     * @param toWorld       The world the player is teleporting to.
+     * @param result        The result of the teleportation.
+     */
+    private void tellReason(@NotNull CommandSender sender,
+                            @NotNull Player teleportee,
+                            @Nullable MVWorld fromWorld,
+                            @Nullable MVWorld toWorld,
+                            @NotNull ActionResult result
+    ) {
+        BukkitCommandIssuer issuer = this.plugin.getMVCommandManager().getCommandIssuer(sender);
+        String targetName = issuer.getIssuer() == teleportee ? "You" : teleportee.getName();
+
+        if (result.hasResult(NullPlaceResult.NULL_DESTINATION)) {
+            issuer.sendMessage(targetName + " cannot be teleported to because the destination is null.");
+            return;
+        }
+        if (result.hasResult(NullPlaceResult.NULL_LOCATION)) {
+            issuer.sendMessage(targetName + " cannot be teleported to because the location is null.");
+            return;
+        }
+        if (result.hasResult(NullPlaceResult.NULL_WORLD)) {
+            issuer.sendMessage(targetName + " cannot be teleported because the world is null.");
```

Comment:
Will this happen when the world is not managed by MV?

Reply:
hmm right... its suppose to be `NOT_MV_WORLD` and returns successful

Reply:
Not quite sure I understand this response... I might be missing the context at this point D:

Reply:
`NULL_WORLD` is when the bukkit world object is null, if world is just not managed by MV, it should be `NOT_MV_WORLD`, which does this:
![image](https://user-images.githubusercontent.com/30431861/221901398-0a885a06-f54e-4c7c-86a0-374ea2e54b7e.png)

which is the same behaviour as before, just refactored

Reply:
I see. Would a name like `NOT_BUKKIT_WORLD` or something make sense?

Reply:
Uh if it's not a bukkit world, means it's null, there isn't such a thing as a non-bukkit world. And then MVWorld is kinda just an extension of bukkit world